File: packages/guards/src/ban-substring.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface BanSubstringOptions {
  action: 'block' | 'warn';
  substrings: string[];
  caseSensitive?: boolean;
  containsAll?: boolean;
}

export function banSubstring(options: BanSubstringOptions): Guard {
  const ci = !(options.caseSensitive ?? false);
  const containsAll = options.containsAll ?? false;

  return {
    name: 'ban-substring',
    version: '0.1.0',
    description: 'Ban specific substrings from text',
    category: 'content',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const searchText = ci ? text.toLowerCase() : text;
      const matched: string[] = [];

      for (const sub of options.substrings) {
        const searchSub = ci ? sub.toLowerCase() : sub;
        if (searchText.includes(searchSub)) matched.push(sub);
      }

      const triggered = containsAll
        ? matched.length === options.substrings.length
        : matched.length > 0;

      return {
        guardName: 'ban-substring',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/ban_substring.py
"""Ban specific substrings from text."""

import time
from typing import List

from open_guardrail.core import GuardResult


class _BanSubstring:
    def __init__(
        self,
        *,
        action: str = "block",
        substrings: List[str],
        case_sensitive: bool = False,
        contains_all: bool = False,
    ) -> None:
        self.name = "ban-substring"
        self.action = action
        self._substrings = substrings
        self._case_sensitive = case_sensitive
        self._contains_all = contains_all

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        search_text = (
            text
            if self._case_sensitive
            else text.lower()
        )
        matched: list[str] = []

        for sub in self._substrings:
            search_sub = (
                sub
                if self._case_sensitive
                else sub.lower()
            )
            if search_sub in search_text:
                matched.append(sub)

        if self._contains_all:
            triggered = (
                len(matched) == len(self._substrings)
            )
        else:
            triggered = len(matched) > 0

        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="ban-substring",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Banned substring found: "
                + ", ".join(matched)
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "matched": matched,
                    "reason": (
                        "Text contains banned"
                        " substrings"
                    ),
                }
                if triggered
                else None
            ),
        )


def ban_substring(
    *,
    action: str = "block",
    substrings: List[str],
    case_sensitive: bool = False,
    contains_all: bool = False,
) -> _BanSubstring:
    return _BanSubstring(
        action=action,
        substrings=substrings,
        case_sensitive=case_sensitive,
        contains_all=contains_all,
    )
